### ISSUE TITLE
Component helper with positionalParams broken

### ISSUE THREAD

**Issue by Emrvb** [emberjs/ember.js#12762]
https://ember-twiddle.com/7a257d88c6cf0efe2b53

Exception "Uncaught Error: Assertion Failed: You cannot specify both a positional param (at position 0) and the hash argument `myParam`."

**Comment by Emrvb** [comment]
Broken since 2.1

I assume by this commit: 8a163a052f2b1aee6fc2fc650d421d7537596724 ([BUGFIX beta] Change preference in positional parameters)

**Comment by Serabe** [comment]
I think I solved a similar issue for closure components. I'll take a look.

**Comment by Emrvb** [comment]
@Serabe 

Yeah, I came across that while searching before submitting this issue. I'm not that familiar with the 2.x codebase. If the param binding in 2.x for regular components hasn't changed much from 1.13 it might be a little harder to solve.

**Comment by diogomafra** [comment]
@Serabe have you found the issue? 

I got the same error when testing the library ember-easy-form against Ember 2.3 beta 3.

**Comment by Serabe** [comment]
The issue is a combination of using `{{component}}` helper, using positional parameters, having the same positional parameter name in both components, and Ember modifying the hash attribute in place.

The rendered path is as follows:

1. Ember finds the component helper and starts the normal render of the first child component.
2. During the first render, positional parameters are processed into the hash and this is modified in place.
3. Rerenders avoid processing positional parameters, therefore there is no problem with those.
4. When the component name changes, the component is rendered (not re-rendered), causing the positional parameters to be processed again and finding a conflict.

There are several ways to fix this, though I might need @mixonic and/or @rwjblue to check which one to follow:

1. Substitute the assertion by a warning.
2. Create a copy of `hash` in the component keyword before passing down to `@element_component`.
3. Refactor the rendering process not to depend on `hash` being modified in place.

**Comment by Serabe** [comment]
@diogomafra your issue is related to [this line](https://github.com/emberjs/ember.js/blob/master/packages/ember-htmlbars/lib/hooks/component.js#L47) and was fixed one day after 2.3 beta3 was released.

**Comment by diogomafra** [comment]
I was trying to create a test replicating my issue but was not able to. Maybe it's because it's already fixed.

I created a test that replicates the issue reported by @Emrvb if that helps.

```js
// packages/ember-htmlbars/tests/helpers/component_test.js

QUnit.test('AAAAA', function() {
  owner.register('component:normal-message', Component.extend({
    something: null
  }).reopenClass({
    positionalParams: ['something']
  }));

  owner.register('component:alternative-message', Component.extend({
    something: null
  }).reopenClass({
    positionalParams: ['something']
  }));

  owner.register('template:components/normal-message', compile('Say: {{something}}!'));
  owner.register('template:components/alternative-message', compile('---: {{something}}!'));

  view = EmberView.create({
    [OWNER]: owner,
    messageType: 'normal-message',
    message: 'Hello',
    template: compile('{{component view.messageType view.message}}')
  });

  runAppend(view);
  equal(view.$().text(), 'Say: Hello!');

  run(function() {
    set(view, 'messageType', 'alternative-message');
  });

  equal(view.$().text(), '---: Hello!');
});
```

**Comment by Serabe** [comment]
Thank you! That saves me time for sure when the course of action is chosen! :+1:

**Comment by pixelhandler** [comment]
@Emrvb does your ember-twiddle example work in v2.0.2 ?

**Comment by Emrvb** [comment]
@Serabe

First I would suggest a warning or deprecation while this issue exists and at least consider backporting that to stable.

The fix you implement should probably work with a copy of the hash. If not, positionalParams processed remain bound to a certain component attribute. This will unintentionally overwrite an attribute on the dynamic component rendered. Which you don't want and what may even be the reason why you were using positionalParams in the first place.

@diogomafra 

My twiddle is still broken when using canary. I also assume a fix would reference this issue.

@pixelhandler 

Yes it works in v2.0.2. I linked the the commit causing the issue which I believe was merged in 2.1.0.

**Comment by Emrvb** [comment]
@diogomafra 

I reviewed your test. Your components both bind the 1st positional param to "something". To trigger the error, the alternative-message should bind its 1st param to "somethingelse".

**Comment by diogomafra** [comment]
@Emrvb The problem I had was not exactly the same as yours. The problem I had is fixed on canary, but the problem you have is not.

The test above is failing. As you can see, it only replaces the component name. I think it is simulating your problem, if not, feel free to change.

**Comment by Emrvb** [comment]
@diogomafra 

Oh I see now you mentioned another issue. Disregard my comment ;)

**Comment by Serabe** [comment]
This is fixed in 2.4.0 for sure. [Twiddle](https://ember-twiddle.com/008315da0ee4212248ede9cafd0c6d06?openFiles=twiddle.json%2C).

Guess this can be closed.

**Comment by rwjblue** [comment]
Thanks @Serabe!
